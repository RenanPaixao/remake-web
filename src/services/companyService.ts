import { supabase } from '../utils/supabase.ts'
import { Location, LocationsService } from './locationService.ts'

export interface Company {
  id: string
  name: string
  created_at?: string
  owner_id: string
}

export interface CompanyWithLocations extends Company {
  locations: Location[]
}

class PlacesServiceImp {
  get queryBuilder() {
    return supabase.from('companies')
  }
  async getAllWithLocations(options?: {from:number, to:number}): Promise<{ count: number, companies: CompanyWithLocations[] }> {
    const { data: companies, count } = options ?
      await this.queryBuilder.select('*, locations(*)', { count: 'exact' }).range(options.from, options.to).throwOnError() :
      await this.queryBuilder.select('*, locations(*)', { count: 'exact' }).throwOnError()

    if (companies === null) {
      return {
        companies: [],
        count: 0
      }
    }

    return {
      companies: companies.filter(company => company.locations.length),
      count: count ?? 0
    }
  }

  async searchCompanies(query: string): Promise<{ count: number, companies: CompanyWithLocations[] }> {
    const { data: companies, count } = await this.queryBuilder
      .select('*, locations(*)', { count: 'exact' })
      .textSearch('name', `${query}`, {
        type: 'websearch'
      })
      .throwOnError()

    if (companies === null) {
      return {
        companies: [],
        count: 0
      }
    }

    return {
      companies: companies.filter(company => company.locations.length),
      count: count ?? 0
    }
  }

  async deleteCompany(id: string): Promise<void> {
    await this.queryBuilder.delete().eq('id', id).throwOnError()
  }

  async createCompanyWithLocation(
    company: Omit<Company, 'id'>,
    location: Omit<Location, 'company_id'>): Promise<void> {
    const { data, error } = await this.queryBuilder.insert(company).select().maybeSingle()

    if(error) {
      throw new Error(error.message)
    }

    try {
      await LocationsService.create({
        ...location,
        company_id: data.id
      })
    }catch(e: any) {
      await this.deleteCompany(data.id)

      if(e.message) {
        throw new Error(e.message)
      }
      throw new Error('Error creating company')
    }
  }

  async getUserPlaces(userId: string): Promise<CompanyWithLocations[]> {
    const { data: companies } = await this.queryBuilder
      .select('*, locations(*)')
      .eq('owner_id', userId)
      .throwOnError()

    if (companies === null) {
      return []
    }

    return companies.filter(company => company.locations.length)
  }
}

export const CompanyService = new PlacesServiceImp()
