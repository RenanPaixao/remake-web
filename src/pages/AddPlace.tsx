import {
  Button,
  Center, Flex,
  FormControl,
  FormErrorMessage, FormHelperText,
  FormLabel, Heading,
  Input,
  SimpleGrid
} from '@chakra-ui/react'
import { useFormik } from 'formik'
import * as Yup from 'yup'
import { hasFormikError } from '../utils/hasFormikError.ts'
import { useNavigate } from 'react-router-dom'
import { FaArrowLeft } from 'react-icons/fa'
import { FaPlus } from 'react-icons/fa6'
import { title } from 'radash'
import { useContext, useState } from 'react'
import { CompanyService } from '../services/companyService.ts'
import { UserContext } from '../context/UserContext.tsx'
import { useSuccessToast } from '../hooks/toast/useSuccessToast.tsx'

const validationSchema = Yup.object({
  company_name: Yup.string().required('Company name is required'),
  cep: Yup.string().required('CEP is required').matches(/^\d{5}-\d{3}$/, 'CEP must be in the format 00000-000'),
  street: Yup.string().required('Street is required'),
  state: Yup.string().required('State is required'),
  number: Yup.string().required('Number is required'),
  complement: Yup.string(),
  city: Yup.string().required('City is required'),
  district: Yup.string().required('District is required'),
  openning_hour: Yup.string().required('Openning hour is required').matches(/^\d{2}:\d{2}$/, 'The format should be 00:00'),
  closing_hour: Yup.string().required('Closing hour is required').matches(/^\d{2}:\d{2}$/, 'The format should be 00:00'),
  latitude: Yup.string().required('Latitude is required').matches(/^-?([0-8]?[0-9]|90)(\.[0-9]{1,10})$/, 'Incorrect format for latitude'),
  longitude: Yup.string().required('Longitude is required').matches(/^-?([0-9]{1,2}|1[0-7][0-9]|180)(\.[0-9]{1,10})$/, 'Incorrect format for longitude')
})

type FormValues = Yup.InferType<typeof validationSchema>

const initialValues: FormValues = {
  company_name: '',
  cep: '',
  street: '',
  state: '',
  number: '',
  complement: '',
  city: '',
  district: '',
  openning_hour: '',
  closing_hour: '',
  latitude: '',
  longitude: ''
}

export const AddPlace = () => {
  const { userInformation } = useContext(UserContext)
  const [isLoading, setIsLoading] = useState(false)

  const navigate = useNavigate()
  const successToast = useSuccessToast()
  const formik = useFormik<FormValues>({
    initialValues,
    validationSchema,
    onSubmit: async values => {
      try{
        if(userInformation.id === null) {
          navigate('/login')
          return
        }

        setIsLoading(true)
        await CompanyService.createCompanyWithLocation({
          name: values.company_name,
          owner_id: userInformation.id
        }, {
          cep: values.cep,
          name: values.street,
          state: values.state,
          number: values.number,
          complement: values.complement,
          city: values.city,
          district: values.district,
          openning_hour: values.openning_hour,
          closing_hour: values.closing_hour,
          latitude: parseFloat(values.latitude),
          longitude: parseFloat(values.longitude)
        })

        successToast({ description: 'Place added successfully' })
        navigate('/account')
      }catch(e) {
        console.error(e)
      }finally{
        setIsLoading(false)
      }
    },
    validateOnBlur: true
  })

  // We need to cast to another type because object.keys returns string[]
  const fieldNames = Object.keys(initialValues) as (keyof FormValues)[]
  const requiredFields = fieldNames.filter(key => key !== 'complement')

  return <Center pt={4} flexDirection={'column'}>
    <Heading py={16}>Add Collect Place</Heading>
    <form onSubmit={formik.handleSubmit}>
      <SimpleGrid columns={2} spacing={4}>
        {
          fieldNames.map(key => {
            return <FormControl
              isDisabled={isLoading}
              key={key}
              mt={4}
              isInvalid={hasFormikError(key, formik.touched, formik.errors)}
              isRequired={requiredFields.some(required => key === required)}
            >
              <FormLabel htmlFor={key}>{title(key)}</FormLabel>
              <Input
                isDisabled={isLoading}
                name={key}
                aria-label={title(key)}
                onChange={formik.handleChange}
                value={formik.values[key]}
                onBlur={formik.handleBlur}
              />
              <FormErrorMessage>{formik.errors[key]}</FormErrorMessage>
            </FormControl>
          })
        }
      </SimpleGrid>
      <FormControl>
        <FormHelperText>You can get the latitude and longitude searching for the place address in google maps</FormHelperText>
      </FormControl>
      <SimpleGrid columns={2} py={14} gap={4}>
        <Flex justifyContent={'end'}>
          <Button
            variant={'outline'}
            isLoading={isLoading}
            isDisabled={isLoading}
            colorScheme={'blue'}
            onClick={() => navigate('/account')}
            leftIcon={<FaArrowLeft/>}>
            Account
          </Button>
        </Flex>
        <Flex justifyContent={'start'}>
          <Button colorScheme={'blue'}
            isLoading={isLoading}
            isDisabled={isLoading || !formik.isValid || !formik.dirty}
            rightIcon={<FaPlus/>}
            onClick={() => formik.handleSubmit()}
          >
            Add Place
          </Button>
        </Flex>
      </SimpleGrid>
    </form>
  </Center>
}
