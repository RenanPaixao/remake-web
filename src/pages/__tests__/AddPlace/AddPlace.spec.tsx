import { screen, waitFor } from '@testing-library/dom'
import { expect } from 'vitest'
import { renderWithStore, user } from '../../../../tests/test-utils.tsx'
import { AddPlace } from '../../AddPlace.tsx'
import { CompanyService } from '../../../services/companyService.ts'

const createCompanyMock = vi.fn()
CompanyService.createCompanyWithLocation = createCompanyMock
const successToastMock = vi.fn()
const useNavigateMock = vi.fn()

describe('AddPlace', () => {
  beforeEach(() => {
    vi.restoreAllMocks()
    vi.mock('react-router-dom', async importOriginal => {
      const actual: object = await importOriginal()
      return {
        ...actual,
        useNavigate: () => useNavigateMock
      }
    })

    vi.mock('../../../hooks/toast/useSuccessToast.tsx',  () => {
      return {
        useSuccessToast: () => successToastMock
      }
    })
  })

  it('should render', async () => {
    const { container } = renderWithStore(<AddPlace/>)

    expect(container).toMatchSnapshot()
  })

  it('should be possible to fill all inputs', async () => {
    const { getByRole } = renderWithStore(<AddPlace/>)
    const addPlaceButton = getByRole('button', { name: /Add collect point/i })

    await fillForm()

    await waitFor(()=> {
      expect(addPlaceButton).toBeEnabled()
    })

    await user.click(addPlaceButton)

    await waitFor(()=> {
      expect(createCompanyMock).toHaveBeenCalledTimes(1)
    })

    expect(successToastMock).toHaveBeenCalledTimes(1)
    expect(useNavigateMock).toHaveBeenCalledTimes(1)
    expect(useNavigateMock).toHaveBeenCalledWith('/account')
  })

  it('should back to account page successfully', async () => {
    const { getByRole } = renderWithStore(<AddPlace/>)

    const accountButton = getByRole('button', { name: /account/i })

    await user.click(accountButton)

    expect(useNavigateMock).toHaveBeenCalledTimes(1)
    expect(useNavigateMock).toHaveBeenCalledWith('/account')
  })

  it('should redirect to login case the user id is not present', async () => {
    const { getByRole } = renderWithStore(<AddPlace/>, {
      user: {
        isAuthenticated: false,
        userInformation: {
          id: null,
          fullName: '',
          email: '',
          isRecycler: false
        }
      }
    })
    const addPlaceButton = getByRole('button', { name: /Add collect point/i })

    await fillForm()

    await waitFor(()=> {
      expect(addPlaceButton).toBeEnabled()
    })

    await user.click(addPlaceButton)

    expect(useNavigateMock).toHaveBeenCalledWith('/login')
    expect(createCompanyMock).not.toHaveBeenCalled()
  })
})

/**
 * Fills place form.
 */
async function fillForm() {
  const { getByRole } = screen

  const companyNameInput = getByRole('textbox', { name: /Company Name/i })
  const cepInput = getByRole('textbox', { name: /Cep/i })
  const streetInput = getByRole('textbox', { name: /Street/i })
  const stateInput = getByRole('textbox', { name: /State/i })
  const numberInput = getByRole('textbox', { name: /Number/i })
  const complementInput = getByRole('textbox', { name: /complement/i })
  const cityInput = getByRole('textbox', { name: /city/i })
  const districtInput = getByRole('textbox', { name: /district/i })
  const openingHourInput = getByRole('textbox', { name: /openning hour/i })
  const closingHourInput = getByRole('textbox', { name: /closing hour/i })
  const latitudeInput = getByRole('textbox', { name: /latitude/i })
  const longitudeInput = getByRole('textbox', { name: /longitude/i })

  await user.type(companyNameInput, 'Company Test')
  await user.type(cepInput, '00000-000')
  await user.type(streetInput, 'street test')
  await user.type(stateInput, 'state test')
  await user.type(numberInput, '18')
  await user.type(complementInput, 'complement test')
  await user.type(cityInput, 'city test')
  await user.type(districtInput, 'district test')
  await user.type(openingHourInput, '08:00')
  await user.type(closingHourInput, '20:00')
  await user.type(latitudeInput, '11.11')
  await user.type(longitudeInput, '11.12')
}
