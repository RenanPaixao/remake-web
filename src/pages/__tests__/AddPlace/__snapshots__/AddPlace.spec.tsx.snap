// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddPlace > should render 1`] = `
<div>
  <div
    class="css-4j83h5"
  >
    <h2
      class="chakra-heading css-1wh6evl"
    >
      Add Collect Place
    </h2>
    <form>
      <div
        class="css-14i3gmy"
      >
        <div
          class="chakra-form-control css-1mn9ptq"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="company_name"
            id="field-:r0:-label"
          >
            Company Name
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1tfjd1n"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-label="Company Name"
            aria-required="true"
            class="chakra-input css-1cjy4zv"
            id="field-:r0:"
            name="company_name"
            required=""
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1mn9ptq"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="cep"
            id="field-:r1:-label"
          >
            Cep
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1tfjd1n"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-label="Cep"
            aria-required="true"
            class="chakra-input css-1cjy4zv"
            id="field-:r1:"
            name="cep"
            required=""
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1mn9ptq"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="street"
            id="field-:r2:-label"
          >
            Street
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1tfjd1n"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-label="Street"
            aria-required="true"
            class="chakra-input css-1cjy4zv"
            id="field-:r2:"
            name="street"
            required=""
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1mn9ptq"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="state"
            id="field-:r3:-label"
          >
            State
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1tfjd1n"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-label="State"
            aria-required="true"
            class="chakra-input css-1cjy4zv"
            id="field-:r3:"
            name="state"
            required=""
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1mn9ptq"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="number"
            id="field-:r4:-label"
          >
            Number
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1tfjd1n"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-label="Number"
            aria-required="true"
            class="chakra-input css-1cjy4zv"
            id="field-:r4:"
            name="number"
            required=""
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1mn9ptq"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="complement"
            id="field-:r5:-label"
          >
            Complement
          </label>
          <input
            aria-label="Complement"
            class="chakra-input css-1cjy4zv"
            id="field-:r5:"
            name="complement"
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1mn9ptq"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="city"
            id="field-:r6:-label"
          >
            City
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1tfjd1n"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-label="City"
            aria-required="true"
            class="chakra-input css-1cjy4zv"
            id="field-:r6:"
            name="city"
            required=""
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1mn9ptq"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="district"
            id="field-:r7:-label"
          >
            District
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1tfjd1n"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-label="District"
            aria-required="true"
            class="chakra-input css-1cjy4zv"
            id="field-:r7:"
            name="district"
            required=""
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1mn9ptq"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="openning_hour"
            id="field-:r8:-label"
          >
            Openning Hour
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1tfjd1n"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-label="Openning Hour"
            aria-required="true"
            class="chakra-input css-1cjy4zv"
            id="field-:r8:"
            name="openning_hour"
            required=""
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1mn9ptq"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="closing_hour"
            id="field-:r9:-label"
          >
            Closing Hour
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1tfjd1n"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-label="Closing Hour"
            aria-required="true"
            class="chakra-input css-1cjy4zv"
            id="field-:r9:"
            name="closing_hour"
            required=""
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1mn9ptq"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="latitude"
            id="field-:ra:-label"
          >
            Latitude
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1tfjd1n"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-label="Latitude"
            aria-required="true"
            class="chakra-input css-1cjy4zv"
            id="field-:ra:"
            name="latitude"
            required=""
            value=""
          />
        </div>
        <div
          class="chakra-form-control css-1mn9ptq"
          role="group"
        >
          <label
            class="chakra-form__label css-g6pte"
            for="longitude"
            id="field-:rb:-label"
          >
            Longitude
            <span
              aria-hidden="true"
              class="chakra-form__required-indicator css-1tfjd1n"
              role="presentation"
            >
              *
            </span>
          </label>
          <input
            aria-label="Longitude"
            aria-required="true"
            class="chakra-input css-1cjy4zv"
            id="field-:rb:"
            name="longitude"
            required=""
            value=""
          />
        </div>
      </div>
      <div
        class="chakra-form-control css-1kxonj9"
        role="group"
      >
        <div
          class="chakra-form__helper-text css-186pyma"
          id="field-:rc:-helptext"
        >
          You can get the latitude and longitude searching for the place address in google maps
        </div>
      </div>
      <div
        class="css-1dcc61u"
      >
        <div
          class="css-ckyhk3"
        >
          <button
            class="chakra-button css-pejez8"
            type="button"
          >
            <span
              class="chakra-button__icon css-1wh2kri"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                focusable="false"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
                />
              </svg>
            </span>
            Account
          </button>
        </div>
        <div
          class="css-8owaep"
        >
          <button
            class="chakra-button css-jut409"
            disabled=""
            type="button"
          >
            Add Place
            <span
              class="chakra-button__icon css-1hzyiq5"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                focusable="false"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </form>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
